File: src/com/ex/stc/Platform.java
/* ========================================================================== */
/*                          STC - SIMPLE TETRIS CLONE                         */
/* -------------------------------------------------------------------------- */
/*   J2ME implementation.                                                     */
/*   CLDC 1.1 - MIDP 2.0 (176 x 220 screen size)                              */
/*   In memorial of my Sony Ericsson W810i, my best phone ever.               */
/*                                                                            */
/*   Copyright (c) 2010 Laurens Rodriguez Oscanoa.                            */
/*   This code is licensed under the MIT license:                             */
/*   http://www.opensource.org/licenses/mit-license.php                       */
/* -------------------------------------------------------------------------- */

package com.ex.stc;

import java.io.IOException;
import java.util.Random;
import javax.microedition.lcdui.Canvas;
import javax.microedition.lcdui.Graphics;
import javax.microedition.lcdui.Image;
import javax.microedition.lcdui.game.Sprite;

public class Platform extends Canvas implements Runnable {
    /*
     * UI layout (quantities are expressed in pixels)
     */

    /* Screen size */
    private static final int SCREEN_WIDTH   = 176;
    private static final int SCREEN_HEIGHT  = 220;

    /* Size of square tile */
    private static final int TILE_SIZE      = 10;

    /* Board up-left corner coordinates */
    private static final int BOARD_X0        = 0;
    private static final int BOARD_Y0        = -1;
    private static final int BOARD_X1        = BOARD_X0 + TILE_SIZE * Game.BOARD_WIDTH;
    private static final int BOARD_Y1        = BOARD_Y0 + TILE_SIZE * Game.BOARD_HEIGHT;

    /* Preview tetromino position */
    private static final int PREVIEW_X      = 127;
    private static final int PREVIEW_Y      = 60;

    /* Score position and length on screen */
    private static final int SCORE_X        = 88;
    private static final int SCORE_Y        = 2;
    private static final int SCORE_LENGTH   = 10;

    /* Lines position and length on screen */
    private static final int LINES_X        = 123;
    private static final int LINES_Y        = 17;
    private static final int LINES_LENGTH   = 5;

    /* Level position and length on screen */
    private static final int LEVEL_X        = 123;
    private static final int LEVEL_Y        = 32;
    private static final int LEVEL_LENGTH   = 5;

    /* Tetromino subtotals position */
    private static final int TETROMINO_X    = 123;
    private static final int TETROMINO_L_Y  = 100;
    private static final int TETROMINO_I_Y  = 115;
    private static final int TETROMINO_T_Y  = 130;
    private static final int TETROMINO_S_Y  = 145;
    private static final int TETROMINO_Z_Y  = 160;
    private static final int TETROMINO_O_Y  = 175;
    private static final int TETROMINO_J_Y  = 190;
    private static final int TETROMINO_LENGTH   = 5;

    /* Tetromino total position */
    private static final int PIECES_X       = 116;
    private static final int PIECES_Y       = 205;
    private static final int PIECES_LENGTH  = 6;

    /* Size of number */
    private static final int NUMBER_WIDTH   = 7;
    private static final int NUMBER_HEIGHT  = 9;

    /* Colors */
    private static final int BACK_COLOR     = 0x424242;
    private static final int GRID_COLOR_LO  = 0x585858;
    private static final int GRID_COLOR_HI  = 0x818181;

    /*
     * Image files
     */
    private static final String BMP_TILE_BLOCKS = "/blocks.png";
    private static final String BMP_NUMBERS     = "/numbers.png";

    /* Delayed autoshift initial delay */
    private static final int DAS_DELAY_TIMER    = 200;

    /* Delayed autoshift timer for left and right moves */
    private static final int DAS_MOVE_TIMER     = 40;

    /* Rotation auto-repeat delay */
    private static final int ROTATION_AUTOREPEAT_DELAY = 375;

    /* Rotation autorepeat timer */
    private static final int ROTATION_AUTOREPEAT_TIMER = 200;

    /* Game instance */
    private Game mGame;

    /* Images of tile blocks and numbers */
    private Image  mImageTiles;
    private Image  mImageNumbers;

    private Random mRandomGenerator;

    /* For delayed autoshift: http://tetris.wikia.com/wiki/DAS */
    private int delayLeft;
    private int delayRight;
    private int delayDown;
    private int delayRotation;
    private long lastTime;

    /*
     * Initializes platform, if there are no problems returns GAME_ERROR_NONE.
     */
    public Platform() {
        /* Create game */
        mGame = new Game();
        mGame.init(this);

        /* Load images for blocks and numbers */
        try {
            mImageTiles = Image.createImage(BMP_TILE_BLOCKS);
            mImageNumbers = Image.createImage(BMP_NUMBERS);
        }
        catch(IOException ioexception) {
            System.out.println("Error loading images: " + ioexception);
        }

        /* Initialize delayed autoshift */
        lastTime = System.currentTimeMillis();
        delayLeft = -1;
        delayRight = -1;
        delayDown = -1;
        if (Game.STC_AUTO_ROTATION) {
            delayRotation = -1;
        }
    }

    /*
     * Release platform allocated resources
     */
    public void end() {
        mGame.end();
        mGame = null;
    }

    /* Return the current system time in milliseconds */
    public long getSystemTime() {
        return System.currentTimeMillis();
    }

    /* Initialize the random number generator */
    public void seedRandom(long seed) {
        mRandomGenerator = new Random();
    }

    /* Return a random positive integer number */
    public int random() {
        return Math.abs(mRandomGenerator.nextInt());
    }

    private void drawTile(Graphics graphics, int x, int y, int tile, boolean shadow) {
        int sourceX = TILE_SIZE * tile;
        int sourceY = shadow? TILE_SIZE : 0;
        graphics.drawRegion(mImageTiles, sourceX, sourceY,
                            TILE_SIZE, TILE_SIZE, Sprite.TRANS_NONE, x, y, 0);
    }

    private void drawNumber(Graphics graphics, int x, int y, long number, int length, int color) {
        int sourceX, destineX;
        int sourceY = NUMBER_HEIGHT * color;
        int pos = 0;
        do {
            destineX  = x + NUMBER_WIDTH * (length - pos);
            sourceX = NUMBER_WIDTH * (int)(number % 10);

            graphics.drawRegion(mImageNumbers, sourceX, sourceY,
                                NUMBER_WIDTH, NUMBER_HEIGHT, Sprite.TRANS_NONE, destineX, y, 0);
            number /= 10;
        } while (++pos < length);
    }

    /*
     * Render the state of the game using platform functions.
     */
    public void paint(Graphics graphics) {
        int i, j;

        /* Check if the game state has changed, if so redraw */
        if (mGame.stateChanged) {

            /* Draw background. */
            graphics.setColor(BACK_COLOR);
            graphics.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);

            graphics.setColor(GRID_COLOR_LO);
            for (i = 0; i <= Game.BOARD_WIDTH; ++i) {
                graphics.drawLine(BOARD_X0 + i * TILE_SIZE, BOARD_Y0, BOARD_X0 + i * TILE_SIZE, BOARD_Y1);
            }
            for (i = 0; i <= Game.BOARD_HEIGHT; ++i) {
                graphics.drawLine(BOARD_X0, BOARD_Y0 + i * TILE_SIZE, BOARD_X1, BOARD_Y0 + i * TILE_SIZE);
            }
            graphics.setColor(GRID_COLOR_HI);
            graphics.drawLine(BOARD_X0, BOARD_Y0 + 2 * TILE_SIZE, BOARD_X0, BOARD_Y1);
            graphics.drawLine(BOARD_X1, BOARD_Y0 + 2 * TILE_SIZE, BOARD_X1, BOARD_Y1);
            graphics.drawLine(BOARD_X0, BOARD_Y0 + 2 * TILE_SIZE, BOARD_X1, BOARD_Y0 + 2 * TILE_SIZE);
            graphics.drawLine(BOARD_X0, BOARD_Y1, BOARD_X1, BOARD_Y1);

            /* Draw preview block */
            if (mGame.showPreview) {
                for (i = 0; i < Game.TETROMINO_SIZE; ++i) {
                    for (j = 0; j < Game.TETROMINO_SIZE; ++j) {
                        if (mGame.nextBlock.cells[i][j] != Game.EMPTY_CELL) {
                            drawTile(graphics,
                                     PREVIEW_X + (TILE_SIZE * i),
                                     PREVIEW_Y + (TILE_SIZE * j),
                                     mGame.nextBlock.cells[i][j], false);
                        }
                    }
                }
            }

            if (Game.STC_SHOW_GHOST_PIECE) {
                /* Draw shadow tetromino */
                if (mGame.showShadow && mGame.shadowGap > 0) {
                    for (i = 0; i < Game.TETROMINO_SIZE; ++i) {
                        for (j = 0; j < Game.TETROMINO_SIZE; ++j) {
                            if (mGame.fallingBlock.cells[i][j] != Game.EMPTY_CELL) {
                                drawTile(graphics,
                                         BOARD_X0 + (TILE_SIZE * (mGame.fallingBlock.x + i)),
                                         BOARD_Y0 + (TILE_SIZE * (mGame.fallingBlock.y + mGame.shadowGap + j)),
                                         mGame.fallingBlock.cells[i][j], true);
                            }
                        }
                    }
                }
            }

            /* Draw the cells in the board */
            for (i = 0; i < Game.BOARD_WIDTH; ++i) {
                for (j = 0; j < Game.BOARD_HEIGHT; ++j) {
                    if (mGame.map[i][j] != Game.EMPTY_CELL) {
                        drawTile(graphics,
                                 BOARD_X0 + (TILE_SIZE * i),
                                 BOARD_Y0 + (TILE_SIZE * j),
                                 mGame.map[i][j], false);
                    }
                }
            }

            /* Draw falling tetromino */
            for (i = 0; i < Game.TETROMINO_SIZE; ++i) {
                for (j = 0; j < Game.TETROMINO_SIZE; ++j) {
                    if (mGame.fallingBlock.cells[i][j] != Game.EMPTY_CELL) {
                        drawTile(graphics,
                                 BOARD_X0 + (TILE_SIZE * (mGame.fallingBlock.x + i)),
                                 BOARD_Y0 + (TILE_SIZE * (mGame.fallingBlock.y + j)),
                                 mGame.fallingBlock.cells[i][j], false);
                    }
                }
            }

            /* Draw game statistic data */
            if (!mGame.isPaused) {
                drawNumber(graphics, LEVEL_X, LEVEL_Y, mGame.stats.level, LEVEL_LENGTH, Game.COLOR_WHITE);
                drawNumber(graphics, LINES_X, LINES_Y, mGame.stats.lines, LINES_LENGTH, Game.COLOR_WHITE);
                drawNumber(graphics, SCORE_X, SCORE_Y, mGame.stats.score, SCORE_LENGTH, Game.COLOR_WHITE);

                drawNumber(graphics, TETROMINO_X, TETROMINO_L_Y, mGame.stats.pieces[Game.TETROMINO_L], TETROMINO_LENGTH, Game.COLOR_ORANGE);
                drawNumber(graphics, TETROMINO_X, TETROMINO_I_Y, mGame.stats.pieces[Game.TETROMINO_I], TETROMINO_LENGTH, Game.COLOR_CYAN);
                drawNumber(graphics, TETROMINO_X, TETROMINO_T_Y, mGame.stats.pieces[Game.TETROMINO_T], TETROMINO_LENGTH, Game.COLOR_PURPLE);
                drawNumber(graphics, TETROMINO_X, TETROMINO_S_Y, mGame.stats.pieces[Game.TETROMINO_S], TETROMINO_LENGTH, Game.COLOR_GREEN);
                drawNumber(graphics, TETROMINO_X, TETROMINO_Z_Y, mGame.stats.pieces[Game.TETROMINO_Z], TETROMINO_LENGTH, Game.COLOR_RED);
                drawNumber(graphics, TETROMINO_X, TETROMINO_O_Y, mGame.stats.pieces[Game.TETROMINO_O], TETROMINO_LENGTH, Game.COLOR_YELLOW);
                drawNumber(graphics, TETROMINO_X, TETROMINO_J_Y, mGame.stats.pieces[Game.TETROMINO_J], TETROMINO_LENGTH, Game.COLOR_BLUE);

                drawNumber(graphics, PIECES_X, PIECES_Y, mGame.stats.totalPieces, PIECES_LENGTH, Game.COLOR_WHITE);
            }

            /* Clear the game state */
            mGame.stateChanged = false;
        }
    }

    /* Read input device and notify game */
    public void keyPressed(int keyCode) {
        switch (getGameAction(keyCode)) {
        case DOWN:
            mGame.events |= Game.EVENT_MOVE_DOWN;
            delayDown = DAS_DELAY_TIMER;
            break;
        case UP:
            mGame.events |= Game.EVENT_ROTATE_CW;
            if (Game.STC_AUTO_ROTATION) {
                delayRotation = ROTATION_AUTOREPEAT_DELAY;
            }
            break;
        case LEFT:
            mGame.events |= Game.EVENT_MOVE_LEFT;
            delayLeft = DAS_DELAY_TIMER;
            break;
        case RIGHT:
            mGame.events |= Game.EVENT_MOVE_RIGHT;
            delayRight = DAS_DELAY_TIMER;
            break;
        case FIRE:
            mGame.events |= Game.EVENT_DROP;
            break;
        }
    }

    public void keyReleased(int keyCode) {
        switch (getGameAction(keyCode)) {
        case DOWN:
            delayDown = -1;
            break;
        case LEFT:
            delayLeft = -1;
            break;
        case RIGHT:
            delayRight = -1;
            break;
        case UP:
            if (Game.STC_AUTO_ROTATION) {
                delayRotation = -1;
            }
            break;
        }
    }

    /* Notification handler when canvas is shown */
    protected void showNotify () {
        new Thread(this).start();
    }

    public void run() {
        /* Game loop */
        while (mGame != null) {
            try {
                /* Update game */
                mGame.update();

                /* Request redraw of screen */
                repaint();

                /* Process delayed autoshift */
                long timeNow = System.currentTimeMillis();
                int timeDelta = (int)(timeNow - lastTime);
                if (delayDown > 0) {
                    delayDown -= timeDelta;
                    if (delayDown <= 0) {
                        delayDown = DAS_MOVE_TIMER;
                        mGame.events |= Game.EVENT_MOVE_DOWN;
                    }
                }
                if (delayLeft > 0) {
                    delayLeft -= timeDelta;
                    if (delayLeft <= 0) {
                        delayLeft = DAS_MOVE_TIMER;
                        mGame.events |= Game.EVENT_MOVE_LEFT;
                    }
                }
                else if (delayRight > 0) {
                    delayRight -= timeDelta;
                    if (delayRight <= 0) {
                        delayRight = DAS_MOVE_TIMER;
                        mGame.events |= Game.EVENT_MOVE_RIGHT;
                    }
                }
                if (Game.STC_AUTO_ROTATION) {
                    if (delayRotation > 0) {
                        delayRotation -= timeDelta;
                        if (delayRotation <= 0) {
                            delayRotation = ROTATION_AUTOREPEAT_TIMER;
                            mGame.events |= Game.EVENT_ROTATE_CW;
                        }
                    }
                }
                lastTime = timeNow;

                /* Resting game */
                Thread.sleep(DAS_MOVE_TIMER);
            }
            catch (java.lang.InterruptedException e) {
                /* Ignore */
            }
        }
    }

    /* Necessary in some devices if using full screen mode. */
    protected void sizeChanged(int w, int h) {
        mGame.stateChanged = true;
        repaint();
    }
}

File: source/platform.lua
-- ========================================================================== --
--   Platform implementation.                                                 --
--   Copyright (c) 2011 Laurens Rodriguez Oscanoa.                            --
-- -------------------------------------------------------------------------- --

-- Screen size 
local SCREEN_WIDTH  = 480;
local SCREEN_HEIGHT = 272;

-- Size of square tile 
local TILE_SIZE = 12;

-- Board up-left corner coordinates 
local BOARD_X = 180;
local BOARD_Y = 4;

-- Preview tetromino position 
local PREVIEW_X = 112;
local PREVIEW_Y = 210;

-- Score position and length on screen 
local SCORE_X      = 72;
local SCORE_Y      = 52;
local SCORE_LENGTH = 10;

-- Lines position and length on screen 
local LINES_X      = 108;
local LINES_Y      = 34;
local LINES_LENGTH = 5;

-- Level position and length on screen 
local LEVEL_X      = 108;
local LEVEL_Y      = 16;
local LEVEL_LENGTH = 5;

-- Tetromino subtotals position 
local TETROMINO_X   = 425;
local TETROMINO_L_Y = 53;
local TETROMINO_I_Y = 77;
local TETROMINO_T_Y = 101;
local TETROMINO_S_Y = 125;
local TETROMINO_Z_Y = 149;
local TETROMINO_O_Y = 173;
local TETROMINO_J_Y = 197;

-- Size of subtotals 
local TETROMINO_LENGTH = 5;

-- Tetromino total position 
local PIECES_X      = 418;
local PIECES_Y      = 221;
local PIECES_LENGTH = 6;

-- Size of number 
local NUMBER_WIDTH  = 7;
local NUMBER_HEIGHT = 9;
    
Platform = { 
    m_bmpBackground = nil;
    m_bmpBlocks = nil;
    m_bmpNumbers = nil;    
    
    m_blocks = nil;
    m_numbers = nil;

    m_musicLoop = nil;
    m_musicIntro = nil;
    m_musicMute = nil;
};
    
-- Initializes platform.
function Platform:init()
    -- Initialize random generator
    math.randomseed(os.time());

    -- Load images.
    self.m_bmpBackground = love.graphics.newImage("back.png");
    
    self.m_bmpBlocks = love.graphics.newImage("blocks.png");
    self.m_bmpBlocks:setFilter("nearest", "nearest");    
    local w = self.m_bmpBlocks:getWidth();
    local h = self.m_bmpBlocks:getHeight();

    -- Load music.
	self.m_musicIntro = love.audio.newSource("stc_theme_intro.ogg");
	self.m_musicIntro:setVolume(0.5);
	self.m_musicIntro:play();
	self.m_musicLoop = love.audio.newSource("stc_theme_loop.ogg", "stream");
	self.m_musicLoop:setLooping(true);
	self.m_musicLoop:setVolume(0.5);
	m_musicMute = false;

    -- Create quads for blocks
    self.m_blocks = {};
    for shadow = 0, 1 do
        self.m_blocks[shadow] = {};
        for color = 0, Game.COLORS - 1 do
            self.m_blocks[shadow][color] = love.graphics.newQuad(TILE_SIZE * color, (TILE_SIZE + 1) * shadow,
                                                                 TILE_SIZE + 1, TILE_SIZE + 1, w, h);
        end
    end
    
    self.m_bmpNumbers = love.graphics.newImage("numbers.png");
    self.m_bmpNumbers:setFilter("nearest", "nearest");
    w = self.m_bmpNumbers:getWidth();
    h = self.m_bmpNumbers:getHeight();
    
    -- Create quads for numbers
    self.m_numbers = {};
    for color = 0, Game.COLORS - 1 do
        self.m_numbers[color] = {};
        for digit = 0, 9 do
            self.m_numbers[color][digit] = love.graphics.newQuad(NUMBER_WIDTH * digit, NUMBER_HEIGHT * color,
                                                                 NUMBER_WIDTH, NUMBER_HEIGHT, w, h);
        end
    end
end

-- Process events and notify game.
function Platform:onKeyDown(key)
    if (key == "escape") then
        love.event.push("quit");
    end
    if ((key == "left") or (key == "a")) then
        Game:onEventStart(Game.Event.MOVE_LEFT);
    end
    if ((key == "right") or (key == "d")) then
        Game:onEventStart(Game.Event.MOVE_RIGHT);
    end
    if ((key == "down") or (key == "s")) then
        Game:onEventStart(Game.Event.MOVE_DOWN);
    end
    if ((key == "up") or (key == "w")) then
        Game:onEventStart(Game.Event.ROTATE_CW);
    end
    if (key == " ") then
        Game:onEventStart(Game.Event.DROP);
    end
    if (key == "f5") then
        Game:onEventStart(Game.Event.RESTART);
    end
    if (key == "f1") then
        Game:onEventStart(Game.Event.PAUSE);
    end
    if (key == "f2") then
        Game:onEventStart(Game.Event.SHOW_NEXT);
    end
    if (key == "f3") then
        Game:onEventStart(Game.Event.SHOW_SHADOW);
    end
end

function Platform:onKeyUp(key)
    if ((key == "left") or (key == "a")) then
        Game:onEventEnd(Game.Event.MOVE_LEFT);
    end
    if ((key == "right") or (key == "d")) then
        Game:onEventEnd(Game.Event.MOVE_RIGHT);
    end
    if ((key == "down") or (key == "s")) then
        Game:onEventEnd(Game.Event.MOVE_DOWN);
    end
    if ((key == "up") or (key == "w")) then
        Game:onEventEnd(Game.Event.ROTATE_CW);
    end
    if (key == "f4") then
		if (self.m_musicMute) then
			if (self.m_musicIntro) then
				self.m_musicIntro:resume();
			else
				self.m_musicLoop:resume();
			end
		else
			if (self.m_musicIntro) then
				self.m_musicIntro:pause();
			else
				self.m_musicLoop:pause();
			end
		end
		self.m_musicMute = not self.m_musicMute;
    end
end

-- Draw a tile from a tetromino
function Platform:drawTile(x, y, tile, shadow)
    love.graphics.drawq(self.m_bmpBlocks, self.m_blocks[shadow][tile], x, y);
end

-- Draw a number on the given position
function Platform:drawNumber(x, y, number, length, color)
    local pos = 0;
    repeat
        love.graphics.drawq(self.m_bmpNumbers, self.m_numbers[color][number % 10],
                            x + NUMBER_WIDTH * (length - pos), y);
        number = math.floor(number / 10);
        pos = pos + 1;
    until (pos >= length);
end

-- Render the state of the game using platform functions.
function Platform:renderGame()
    -- Draw background
    love.graphics.draw(self.m_bmpBackground, 0, 0);

    -- Draw preview block
    if Game:showPreview() then
        for i = 0, Game.TETROMINO_SIZE - 1 do
            for j = 0, Game.TETROMINO_SIZE - 1 do
                if (Game:nextBlock().cells[i][j] ~= Game.Cell.EMPTY) then
                    Platform:drawTile(PREVIEW_X + TILE_SIZE * i,
                                      PREVIEW_Y + TILE_SIZE * j,
                                      Game:nextBlock().cells[i][j], 0);
                end
            end
        end
    end

    -- Draw shadow tetromino
    if (Game:showShadow() and Game:shadowGap() > 0) then
        for i = 0, Game.TETROMINO_SIZE - 1 do
            for j = 0, Game.TETROMINO_SIZE - 1 do
                if (Game:fallingBlock().cells[i][j] ~= Game.Cell.EMPTY) then
                    Platform:drawTile(BOARD_X + (TILE_SIZE * (Game:fallingBlock().x + i)),
                                      BOARD_Y + (TILE_SIZE * (Game:fallingBlock().y + Game:shadowGap() + j)),
                                      Game:fallingBlock().cells[i][j], 1);
                end
            end
        end
    end

    -- Draw the cells in the board
    for i = 0, Game.BOARD_TILEMAP_WIDTH - 1 do
        for j = 0, Game.BOARD_TILEMAP_HEIGHT - 1 do
            if (Game:getCell(i, j) ~= Game.Cell.EMPTY) then
                Platform:drawTile(BOARD_X + (TILE_SIZE * i),
                                  BOARD_Y + (TILE_SIZE * j),
                                  Game:getCell(i, j), 0);
            end
        end
    end

    -- Draw falling tetromino
    for i = 0, Game.TETROMINO_SIZE - 1 do
        for j = 0, Game.TETROMINO_SIZE - 1 do
            if (Game:fallingBlock().cells[i][j] ~= Game.Cell.EMPTY) then
                Platform:drawTile(BOARD_X + TILE_SIZE * (Game:fallingBlock().x + i),
                                  BOARD_Y + TILE_SIZE * (Game:fallingBlock().y + j),
                                  Game:fallingBlock().cells[i][j], 0);
            end
        end
    end
    
    -- Draw game statistic data
    if (not Game:isPaused()) then
        Platform:drawNumber(LEVEL_X, LEVEL_Y, Game:stats().level, LEVEL_LENGTH, Game.Cell.WHITE);
        Platform:drawNumber(LINES_X, LINES_Y, Game:stats().lines, LINES_LENGTH, Game.Cell.WHITE);
        Platform:drawNumber(SCORE_X, SCORE_Y, Game:stats().score, SCORE_LENGTH, Game.Cell.WHITE);

        Platform:drawNumber(TETROMINO_X, TETROMINO_L_Y, Game:stats().pieces[Game.TetrominoType.L], TETROMINO_LENGTH, Game.Cell.ORANGE);
        Platform:drawNumber(TETROMINO_X, TETROMINO_I_Y, Game:stats().pieces[Game.TetrominoType.I], TETROMINO_LENGTH, Game.Cell.CYAN);
        Platform:drawNumber(TETROMINO_X, TETROMINO_T_Y, Game:stats().pieces[Game.TetrominoType.T], TETROMINO_LENGTH, Game.Cell.PURPLE);
        Platform:drawNumber(TETROMINO_X, TETROMINO_S_Y, Game:stats().pieces[Game.TetrominoType.S], TETROMINO_LENGTH, Game.Cell.GREEN);
        Platform:drawNumber(TETROMINO_X, TETROMINO_Z_Y, Game:stats().pieces[Game.TetrominoType.Z], TETROMINO_LENGTH, Game.Cell.RED);
        Platform:drawNumber(TETROMINO_X, TETROMINO_O_Y, Game:stats().pieces[Game.TetrominoType.O], TETROMINO_LENGTH, Game.Cell.YELLOW);
        Platform:drawNumber(TETROMINO_X, TETROMINO_J_Y, Game:stats().pieces[Game.TetrominoType.J], TETROMINO_LENGTH, Game.Cell.BLUE);

        Platform:drawNumber(PIECES_X, PIECES_Y, Game:stats().totalPieces, PIECES_LENGTH, Game.Cell.WHITE);
    end

	-- Adding music loop check here for convenience.
	if (self.m_musicIntro) then
		if (self.m_musicIntro:isStopped()) then
			self.m_musicIntro = nil;
			self.m_musicLoop:play();
		end
	end
end

function Platform:getSystemTime()
    return math.floor(1000 * love.timer.getTime());
end

function Platform:random()
    return math.random(1000000000);
end
